app/src/main/java/com/hedvig/app/feature/embark/passages/textactionset/TextActionFragment.kt

```diff
@@ -90,19 +95,61 @@ class TextActionSetFragment : Fragment(R.layout.fragment_text_action_set) {
         delay(PASSAGE_ANIMATION_DELAY_MILLIS)
     }
 
-    private fun textFieldData(data: TextActionSetParameter) =
-        data.keys.mapIndexed { index, key ->
-            TextFieldData(
-                key,
-                data.placeholders[index],
-                data.mask[index],
-                key?.let { model.getFromStore(it) },
-            )
+    private fun createInputViews(): List<View> {
+        return data.keys.mapIndexed { index, key ->
+            val inputView = LayoutInflater.from(requireContext()).inflate(R.layout.embark_input_item, binding.inputContainer, false)
+            val inputLayout = inputView.findViewById<TextInputLayout>(R.id.textField)
+            val inputEditText = inputView.findViewById<TextInputEditText>(R.id.input)
+
+            inputLayout.hint = data.placeholders[index]
+            val mask = data.mask[index]
+            mask?.let {
+                inputEditText.apply {
+                    setInputType(it)
+                    setValidationFormatter(it)
+                }
+            }
+            inputEditText.onChange { text ->
+                if (mask == null) {
+                    if (text.isBlank()) {
+                        textActionSetViewModel.updateIsValid(index, false)
+                    } else {
+                        textActionSetViewModel.updateIsValid(index, true)
+                    }
+                } else {
+                    if (text.isNotBlank() && validationCheck(mask, text)) {
+                        textActionSetViewModel.updateIsValid(index, true)
+                    } else {
+                        textActionSetViewModel.updateIsValid(index, false)
+                    }
+                }
+                textActionSetViewModel.setInputValue(index, text)
+            }
+
+            if (index < data.keys.size - 1) {
+                inputEditText.imeOptions = EditorInfo.IME_ACTION_NEXT
+            } else {
+                inputEditText.imeOptions = EditorInfo.IME_ACTION_DONE
+            }
+
+            inputEditText.onImeAction {
+                if (textActionSetViewModel.isValid.value == true) {
+                    viewLifecycleScope.launch {
```

Comment:
This probably needs to be implemented exactly the same as how the submit-button is implemented, otherwise it will handle multiple clicks on the IME submit improperly.

Reply:
The IME seems to handle multiple clicks really well actually, dont think we need to change this 👍